From 5be7e91cfa396c45feb7a784ebb565d53b2c380c Mon Sep 17 00:00:00 2001
From: Oleksiy Protas <elfy.ua@gmail.com>
Date: Sun, 28 Jul 2024 13:53:23 +0000
Subject: [PATCH] jifmux: Make copies of parsed image buffer parts

Prevents a segmentation fault when the original buffer is freed too fast.
---
 .../gst-plugins-bad/gst/jpegformat/gstjifmux.c       | 12 ++++++++----
 .../gst-plugins-bad/gst/jpegformat/gstjifmux.h       |  2 +-
 2 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/gst/jpegformat/gstjifmux.c b/gst/jpegformat/gstjifmux.c
index 13930874dc..0bbeb1df01 100644
--- a/gst/jpegformat/gstjifmux.c
+++ b/gst/jpegformat/gstjifmux.c
@@ -92,7 +92,7 @@ typedef struct _GstJifMuxMarker
   guint8 marker;
   guint16 size;

-  const guint8 *data;
+  guint8 *data;
   gboolean owned;
 } GstJifMuxMarker;

@@ -250,7 +250,8 @@ gst_jif_mux_new_marker (guint8 marker, guint16 size, const guint8 * data,

   m->marker = marker;
   m->size = size;
-  m->data = data;
+  m->data = g_malloc (size);
+  memcpy (m->data, data, size);
   m->owned = owned;

   return m;
@@ -333,9 +334,10 @@ gst_jif_mux_parse_image (GstJifMux * self, GstBuffer * buf)

       /* remaining size except EOI is scan data */
       self->scan_size = eoi_pos - gst_byte_reader_get_pos (&reader);
-      if (!gst_byte_reader_get_data (&reader, self->scan_size,
-              &self->scan_data))
+      if (!gst_byte_reader_get_data (&reader, self->scan_size, &data))
         goto error;
+      self->scan_data = g_malloc (self->scan_size);
+      memcpy (self->scan_data, data, self->scan_size);

       GST_DEBUG_OBJECT (self, "scan data, size = %u", self->scan_size);
     }
@@ -687,6 +689,8 @@ gst_jif_mux_recombine_image (GstJifMux * self, GstBuffer ** new_buf,
   }
   gst_buffer_unmap (buf, &map);
   gst_byte_writer_free (writer);
+  g_free (self->scan_data);
+  self->scan_data = NULL;

   if (!writer_status) {
     GST_WARNING_OBJECT (self, "Failed to write to buffer, calculated size "
diff --git a/gst/jpegformat/gstjifmux.h b/gst/jpegformat/gstjifmux.h
index e65c873a37..2cb0419e1d 100644
--- a/gst/jpegformat/gstjifmux.h
+++ b/gst/jpegformat/gstjifmux.h
@@ -50,7 +50,7 @@ struct _GstJifMux {
   /* list of GstJifMuxMarker */
   GList *markers;
   guint scan_size;
-  const guint8 *scan_data;
+  guint8 *scan_data;
 };

 struct _GstJifMuxClass {
--
2.48.1

